Makefile

```diff
@@ -50,7 +60,8 @@ tests:
 test/%.bats: ALWAYS_RUN
 	@ ./test/bats/bin/bats $@
 
-lint: eclint hadolint shellcheck
+lint:
+	$(call docker-execute,"[[ ! -e "$(TOOLS_DIR)/hadolint" ]] && make install_linters; make hadolint shellcheck eclint")
```

Comment:
I'm also not that familiar with `make` and it's config, I assume that's why you're not able to run the `docker run` command (that's been nested into `docker-execute`) here directly?

As this wasn't being done before (previously lint commands were called directly, or their makefile entries?), have I understood this correctly that the linting is now being performed inside a docker container?

Is the built `docker-mailserver` image providing some value to do that in vs a more generic container? (as in, was the `lint build` order for `all` swapped to `build lint` just to satisfy this change?)

Reply:
The lint.sh script used by eclint, hadolint, and shellcheck makefile functions do not work on macOS. It (macOS) does not support `realpath` and `readlink` flags that are only available on the GNU versions of those commands.

Secondly, we could use a generic container for lint testing, but I already build the Dockerfile container/tag for testing anyway. The switch of makefile functions `build` and `lint` so that build comes first solves this.

Reply:
Oh, additionally, I defined the docker-execute to make it easier to run other things inside of the built `mailserver-testing:ci` and keep things standardized. I can remove it if you're not liking it.

Reply:
> For example, it does not support the `readlink -f` flag that is only available on the GNU versions of those commands.

:+1: I'm in favor of using docker for running linting :)

> Why clutter my env and use disk space? The switch of makefile functions `build` and `lint` order, so that build comes first, solves this.

You are favoring disk space savings to avoid additional temporary images for running linting over project maintenance? As someone who's primary system is often lacking disk space I can understand wanting to (and often why I tend to prefer small images like Alpine), but at the same time.. I'd rather not introduce new ENV + ARG to the Dockerfile just to support adding a build variant with the lint tools to run.

Each of these lint tools do have an official Docker image. Those can be used for linting in general for various projects which imo is preferable vs each project DIY including their own with approaches like we have been. We don't have to include the new vars for kernel and arch either if that were the case? (since DockerHub takes care of that when pulling images published with different arch implicitly)

I'm not blocking this, it doesn't bother me that much and can always be addressed in future if desired. FWIW, all three official docker images is less than 30MB on disk, which makes the argument in favor difficult to justify imo. Single docker-compose can run all 3 in parallel and does not require keeping versions updated in both a Makefile and our Github Actions workflow.

---

> The switch of makefile functions `build` and `lint` order, so that build comes first, solves this.

I don't see much harm in this, however I still think it's better to pass lints before perfoming a build. Especially within CI where you can have quotas for bandwidth and build/CI minutes (I don't think that's a concern for this project), where failing early is preferable.

If we do adopt linting with official docker images instead, there's no need to prioritizing building first, and this order can be reverted.

---

> I defined the docker-execute to make it easier to run other things inside of the built `mailserver-testing:ci` while keeping it standardized/dry. I can remove it if you're not liking it.

I don't personally see a need or value in it when it's only used once and would be easier to grok without the indirection. I'm not opposed to it, thanks for the clarification.

Reply:
Good points. I agree too. I like the direction of running the lint docker containers and moving lint back before build. I'll try making some changes!

Reply:
Ok! First I want to thank you for such thoughtful feedback and for taking the time to help me out. As you can tell I'm getting used to how things are done in the repo and really look forward to continued contributions and improvements!

Second, I've made what I think is the best way to lint using containers. There are no legit eclint or shellcheck dockerhub tags that have bash available, so I'm just using the same tag we build FROM. I also moved a lot of the logic into lint.sh and out of the Makefile. Not sure if that's ok or not. 

Thanks again!

Reply:
> There are no legit eclint or shellcheck dockerhub tags that have bash available (required by the lint.sh for some reason), so I'm just using the same tag we build FROM.

Well, that's a little better I guess.

> I also moved a lot of the logic into lint.sh and out of the Makefile. Not sure if that's ok or not.

A major reason of wanting to push for the official images was to not require us to need the additional install code in the project :sweat_smile:

Sorry if that wasn't clear, I don't mean to waste your time or disrespect your contributions by dismissing the effort here. I'm just not sure that it's the right approach.

You're welcome to wait on feedback from the other maintainers btw, my maintenance role is minor in this area of the project. The others may have different opinions.

---

For better visibility, I'll share what I had in mind on the main discussion thread.